Library/OcDeviceMiscLib/ForgeUefi.c

```diff
@@ -67,21 +67,43 @@ OcCreateEventEx (
   return EFI_SUCCESS;
 }
 
+//
+// The Trash strategy relies on old Apple EFI allocating gBS and gDS consecutively.
+// This layout is directly inherited from standard edk EFI code.
+// What is being checked for in the Trash strategy is that the QWORD about to be trashed
+// contains DXE_SERVICES_SIGNATURE, a value which happily is only used when the memory is
+// being loaded (when we check for references to this value throughout the edk code).
+// For the Trash strategy to work, we are required to trash exactly that QWORD of memory,
+// but in the targeted firmware we can confirm that it is harmless to do so before proceeding.
+//
 EFI_STATUS
 OcForgeUefiSupport (
-  VOID
+  IN BOOLEAN  Forge,
+  IN BOOLEAN  Trash
   )
 {
   EFI_BOOT_SERVICES  *NewBS;
+  UINT64             Signature;
 
   DEBUG ((
     DEBUG_INFO,
-    "OCDM: Found 0x%X UEFI version (%u bytes, rebuilding to %u)\n",
+    "OCDM: Found 0x%X/0x%X UEFI version (%u bytes, %u %a to %u) gST %p gBS %p gBS->CreateEventEx %p &gBS %p\n",
     gST->Hdr.Revision,
+    gBS->Hdr.Revision,
     gBS->Hdr.HeaderSize,
-    (UINT32)sizeof (EFI_BOOT_SERVICES)
+    Forge,
+    Trash ? "trashing" : "rebuilding",
+    (UINT32)sizeof (EFI_BOOT_SERVICES),
+    gST,
+    gBS,
+    gBS->CreateEventEx,
+    &gBS
     ));
 
+  if (!Forge) {
+    return EFI_SUCCESS;
+  }
+
   //
   // Already too new.
   //
```

Comment:
ForgeUEFI should guard against being passed already forged instances below.

That is, something similar to the `Spoof Proofing` added to OpenVariableRuntimeDxe here: https://github.com/acidanthera/OpenCorePkg/pull/405

Reply:
It already does

Reply:
You mean it will filter this spoof example out:
- SystemTable = UEFI 2.10
- BootServices = EFI 1.10
- RunTimeServices = EFI 1.10

Reply:
Well expand the code below in GitHub and look, I would say clearly, yes, it will filter that. I did not modify these checks.

Reply:
I only see a check on SystemTable against UEFI 2.3 (which would pass) and a check against BootServices size (which is not fool proof) but I suppose I am missing the obvious.

I know you did not change this element btw. Just thought you might want to improve the function given that you are making changes there. Not needed in any case given what you said though.

Reply:
@dakanji, The _immediately_ following lines here are:

```
  if (gST->Hdr.Revision >= EFI_2_30_SYSTEM_TABLE_REVISION) {
    return EFI_ALREADY_STARTED;
  }
```

which means, given the data you mentioned, OcForgeUefiSupport will abort (i.e. not forge anything) immediately, returning EFI_ALREADY_STARTED - which I still make the correct behaviour, no? Or are you saying you _want_ it to forge in the case you have given?

Reply:
I saw that but in the case I posted, gST at 2.10 would not abort here (2.30 and greater abort).

Reply:
Oh dear, you're right, of course, but then I'm still not quite sure what we'd want it to do, in this case.

Are you stating that, as far as you know, the above has example forged elements, e.g. CreateEventEx, within gBS, even though the version of gBS has not been updated?

Reply:
I saw your PM with more detail - the answer to my question is yes - thank you.

We/I do need to process what we want to happen here, though.

Reply:
Mind you, we are still going to not forge, for the firmware you've mentioned, based on the gBS header size, right? (Feel free to set me straight again!)

Reply:
I think testing all three elements and aborting if not all UEFI 2.x is the best course of action.

Btw, it doesn't really matter that in this case, BootServices seems to have had CreateEventEx added.
The mere presence of version inconsistency is the indication of potential issues.

This was the same reason for checking them all in the driver.

Reply:
I now fully understand your reason for [checking them all](https://github.com/acidanthera/OpenCorePkg/pull/405) - thank you. But I'm still not sure checking them all is right...

Maybe we should do something like:

 - Check that we do not already have the minimum UEFI version which would provide the to-be-spoofed item in gST, abort if we do
 - Check that we do not already have the minimum UEFI version which would provide the to-be-spoofed item in gBS (or gRT, when spoofing gRT), abort if we do
 - Check that we do not already have a header size large enough to contain the to-be-spoofed item, in the header of the table we are spoofing (i.e. gBS or gRT, but only the one we are spoofing), abort if we do
 - Othwerwise, spoof (or, as the case is in the emu nvram code, provide in the first place)
 
This would mean some changes to [your check](https://github.com/acidanthera/OpenCorePkg/pull/405) - namely removing the check on gBS version, since there we are spoofing gRT; and some changes to the checks here - namely changing the version check down to 2+, even if we eventually spoof to 2.3.

I'll do a PR of these changes, to get some feedback from yourself and Vit. (Though feel free to feedback here, first.)

Reply:
I don't think the listed changes covers anything that is not already covered or might need to be but will leave things to gurus.

> some changes to the checks here - namely changing the version check down to 2+
I similarly don't think this is as good, or as comprehensive (I can imagine situations that will slip through) as a triple check, but will defer.

Reply:
I think maybe you are thinking only of changes that help when you want to avoid spoofing, where I'm trying to also think of changes that help when you want to avoid NOT spoofing?

Reply:
Anyway, if I do a PR, we can see what you, me, and gurus think!

Reply:
> I think maybe you are thinking only of changes that help when you want to avoid spoofing

Yes. I am indeed thinking that if someone has been spoofing beforehand, then further spoofing should be avoided.
Gurus, such as yourself, may have of course have other considerations!

Reply:
As currently provided, this _will_ replace an earlier forge to 2.0 <= UEFI < 2.3. I believe this is desirable in some cases and harmless in others.